File: tests/test_package_path_is_content_addressed.py
"""A re-uploaded package does not reuse the previous upload's URL.

Every upload of a given version wrote over the last one at a fixed blob path,
and those paths are public CDN URLs. On 2026-09-21 a corrected package was
uploaded and the vetting run that started moments later downloaded the PREVIOUS
code: the report named a traceback from a line the new package does not contain,
and a secret finding nothing in it matches. Re-running minutes later passed 5/5
on the identical upload.

The confusing direction is the harmless one. The other direction is a package
whose new code fails being vetted as the old code that passed, and approved on
that, which is the vetting gate reporting on something other than what ships.

The behaviour is tested in test_package_path_is_content_addressed.mjs so it runs
the shipped function; this file runs it under pytest and checks the upload path
is built from it.
"""
import shutil
import subprocess
from pathlib import Path

import pytest

ROOT = Path(__file__).resolve().parents[1]
STORAGE = (ROOT / "apps" / "web" / "lib" / "package-storage.ts").read_text(encoding="utf-8")
SCRIPT = Path(__file__).resolve().parent / "test_package_path_is_content_addressed.mjs"


@pytest.mark.skipif(shutil.which("node") is None, reason="node is needed")
def test_the_fingerprint_behaves():
    r = subprocess.run(
        ["node", "--experimental-strip-types", str(SCRIPT)],
        capture_output=True, text=True, cwd=SCRIPT.parent,
    )
    print(r.stdout or r.stderr)
    assert r.returncode == 0, r.stdout + r.stderr
    assert r.stdout.count("ok   ") >= 7, r.stdout


def test_the_upload_path_carries_the_fingerprint():
    start = STORAGE.index("export async function storeExtractedPackage")
    body = STORAGE[start:STORAGE.index("\n}\n", start)]
    assert "packageFingerprint(files)" in body, "the prefix must depend on the content"


def test_the_fingerprint_reads_both_names_and_bytes():
    # A package differing only in which file holds which bytes is a different
    # package, and must not share a URL with the other one.
    start = STORAGE.index("export function packageFingerprint")
    body = STORAGE[start:STORAGE.index("\n}\n", start)]
    assert "h.update(relativePath" in body
    assert "files.get(relativePath)" in body

File: tests/test_package_path_is_content_addressed.mjs
/**
 * An upload's blob path depends on what is in the package.
 *
 * Re-running the fingerprint is cheap; the property it protects is not. See the
 * comment on storeExtractedPackage for the run this came from.
 */
import assert from "node:assert";
import { packageFingerprint } from "../apps/web/lib/package-storage.ts";

let passed = 0;
function check(name, fn) {
  fn();
  console.log(`ok   ${name}`);
  passed++;
}

const pkg = (entries) =>
  new Map(Object.entries(entries).map(([k, v]) => [k, Buffer.from(v, "utf8")]));

const BASE = pkg({
  "agent.py": "llm = ChatOpenAI(model=os.environ['LLM_MODEL'])\n",
  "marketplace.json": '{"slug":"action-items"}',
});

check("the same package fingerprints the same", () => {
  assert.strictEqual(packageFingerprint(BASE), packageFingerprint(pkg({
    "agent.py": "llm = ChatOpenAI(model=os.environ['LLM_MODEL'])\n",
    "marketplace.json": '{"slug":"action-items"}',
  })));
});

check("changing a byte of code changes the path", () => {
  // The case that mattered: one line of agent.py corrected, everything else
  // identical, same version number.
  const fixed = pkg({
    "agent.py": "llm = StructuredLLM(model=os.environ['LLM_MODEL'])\n",
    "marketplace.json": '{"slug":"action-items"}',
  });
  assert.notStrictEqual(packageFingerprint(BASE), packageFingerprint(fixed));
});

check("adding a file changes the path", () => {
  const extra = pkg({
    "agent.py": "llm = ChatOpenAI(model=os.environ['LLM_MODEL'])\n",
    "marketplace.json": '{"slug":"action-items"}',
    "requirements.txt": "langchain-openai\n",
  });
  assert.notStrictEqual(packageFingerprint(BASE), packageFingerprint(extra));
});

check("renaming a file changes the path", () => {
  // Same bytes, different name: the package is not the same package.
  const a = pkg({ "agent.py": "x", "b.py": "y" });
  const b = pkg({ "agent.py": "y", "b.py": "x" });
  assert.notStrictEqual(packageFingerprint(a), packageFingerprint(b));
});

check("insertion order does not change the path", () => {
  // Zip readers do not promise an order, and an upload that fingerprinted
  // differently for the same files would defeat the dedupe.
  const a = new Map([["a.py", Buffer.from("1")], ["b.py", Buffer.from("2")]]);
  const b = new Map([["b.py", Buffer.from("2")], ["a.py", Buffer.from("1")]]);
  assert.strictEqual(packageFingerprint(a), packageFingerprint(b));
});

check("a path separator difference does not change the path", () => {
  // The upload normalises backslashes; the fingerprint must agree, or a zip
  // built on Windows would store under a different prefix than the same
  // package built anywhere else.
  const win = new Map([["onboarding\\questions.json", Buffer.from("{}")]]);
  const posix = new Map([["onboarding/questions.json", Buffer.from("{}")]]);
  assert.strictEqual(packageFingerprint(win), packageFingerprint(posix));
});

check("the fingerprint is short enough to live in a path", () => {
  const fp = packageFingerprint(BASE);
  assert.match(fp, /^[0-9a-f]{12}$/);
});

console.log(`\n${passed} checks passed`);
